ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/InputsAndEffects.scala

```diff
@@ -41,64 +60,116 @@ private[kvutils] object InputsAndEffects {
+        }
+      }
     }
 
-    tx.fold(packageInputs: Set[DamlStateKey]) {
-        case (inputs, (nodeId, node)) =>
-          node match {
-            case fetch: NodeFetch[ContractId] =>
-              inputs ++ contractInputs(fetch.coid) ++ partyInputs(fetch.signatories) ++
-                partyInputs(fetch.stakeholders)
-            case create: NodeCreate[ContractId, VersionedValue[ContractId]] =>
-              create.key.fold(inputs) { keyWithM =>
-                inputs + contractKeyToStateKey(
-                  GlobalKey(create.coinst.template, forceNoContractIds(keyWithM.key)))
-              } ++ partyInputs(create.signatories) ++ partyInputs(create.stakeholders)
-            case exe: NodeExercises[_, ContractId, _] =>
-              inputs ++ contractInputs(exe.targetCoid) ++ partyInputs(exe.stakeholders) ++ partyInputs(
-                exe.signatories) ++ partyInputs(exe.controllers)
-            case l: NodeLookupByKey[ContractId, Transaction.Value[ContractId]] =>
-              // We need both the contract key state and the contract state. The latter is used to verify
-              // that the submitter can access the contract.
-              l.result.fold(inputs)(inputs ++ contractInputs(_)) +
-                contractKeyToStateKey(GlobalKey(l.templateId, forceNoContractIds(l.key.key)))
-          }
+    def addPartyInputsInValues(): Unit =
+      tx.foldValues(()) {
+        case (_, VersionedValue(_, v)) =>
+          addPartyInputInValue(v)
       }
-      .toList
+
+    tx.foreach {
+      case (_, node) =>
+        node match {
+          case fetch: NodeFetch[ContractId] =>
+            addContractInput(fetch.coid)
+            inputs ++= partyInputs(fetch.signatories)
+            inputs ++= partyInputs(fetch.stakeholders)
+
+          case create: NodeCreate[ContractId, VersionedValue[ContractId]] =>
+            create.key.foreach { keyWithM =>
+              inputs += contractKeyToStateKey(
+                GlobalKey(create.coinst.template, forceNoContractIds(keyWithM.key)))
+            }
+            inputs ++= partyInputs(create.signatories)
+            inputs ++= partyInputs(create.stakeholders)
+
+          case exe: NodeExercises[_, ContractId, _] =>
+            addContractInput(exe.targetCoid)
+            inputs ++= partyInputs(exe.stakeholders)
+            inputs ++= partyInputs(exe.signatories)
+            inputs ++= partyInputs(exe.controllers)
+
+          case l: NodeLookupByKey[ContractId, Transaction.Value[ContractId]] =>
+            // We need both the contract key state and the contract state. The latter is used to verify
+            // that the submitter can access the contract.
+            l.result.foreach(addContractInput)
+            inputs += contractKeyToStateKey(GlobalKey(l.templateId, forceNoContractIds(l.key.key)))
+        }
+    }
+
+    addPartyInputsInValues()
+
+    inputs.toSet
   }
 
+  def subValues(v: Value[ContractId]): List[Value[ContractId]] =
```

Comment:
is it possible to avoid creating the temporary list and just traverse the value and insert the parties into inputs?

Reply:
It is possible but I decided not to go that far because I might be forced to stop reusing that same function in the validation stage. I can try a bit harder though if you think it's worth the investment.